context/src/main/java/io/opentelemetry/context/propagation/CompositeTextMapPropagator.java

```diff
@@ -15,18 +15,31 @@
 import java.util.Set;
 import javax.annotation.Nullable;
 
-final class MultiTextMapPropagator implements TextMapPropagator {
-  private final TextMapPropagator[] textPropagators;
+final class CompositeTextMapPropagator implements TextMapPropagator {
+  private final TextMapPropagator[] extractors;
+  private final TextMapPropagator[] injectors;
   private final Collection<String> allFields;
 
-  MultiTextMapPropagator(TextMapPropagator... textPropagators) {
+  CompositeTextMapPropagator(TextMapPropagator... textPropagators) {
     this(Arrays.asList(textPropagators));
   }
 
-  MultiTextMapPropagator(List<TextMapPropagator> textPropagators) {
-    this.textPropagators = new TextMapPropagator[textPropagators.size()];
-    textPropagators.toArray(this.textPropagators);
-    this.allFields = Collections.unmodifiableList(getAllFields(this.textPropagators));
+  CompositeTextMapPropagator(List<TextMapPropagator> textPropagators) {
+    this.extractors = new TextMapPropagator[textPropagators.size()];
+    textPropagators.toArray(this.extractors);
+    this.injectors = this.extractors; // No need to copy
+    this.allFields = Collections.unmodifiableList(getAllFields(this.injectors));
+  }
+
+  CompositeTextMapPropagator(
+      List<TextMapPropagator> extractors, List<TextMapPropagator> injectors) {
+    this.extractors = new TextMapPropagator[extractors.size()];
+    extractors.toArray(this.extractors);
+
+    this.injectors = new TextMapPropagator[injectors.size()];
+    injectors.toArray(this.injectors);
+
+    this.allFields = Collections.unmodifiableList(getAllFields(this.injectors));
```

Comment:
is this right? The fields are only from the injectors? I think this might need to be the union of the extractors and injectors, rather than just the injectors.

Reply:
That's not what I understood from the `fields()` javadoc, it rather seems to me that method is used only when writing the context back to the wire, and not while extracting it...

Reply:
I think the javadoc provides that as a usage (to clear out headers if the header implementation is re-used), but I think `fields()` could also be used by extraction in some cases. I'm honestly not 100% sure what the behavior should be in this case when injection doesn't necessarily match extraction.

Reply:
I created a spec issue for this, since I don't think I know what the answer should be:  https://github.com/open-telemetry/opentelemetry-specification/issues/1809

Reply:
I second @jkwatson with this - I know at least one case in instrumentation where `fields` are considered during extraction (see `io.opentelemetry.instrumentation.awslambda.v1_0.TracingRequestStreamHandler`)

Reply:
Let's take a concrete case I have in mind: I have 2 extractors, 1 injector, assuming fields returns the union of the header names used by extractors and injectors, if you use `fields` to preallocate the headers when writing back to the wire, you're gonna waste allocations because you're gonna allocate for those extractors as well. 

Perhaps you need 2 methods for getting the fields, one for injectors and one for extractors?

Reply:
Please comment on the spec issue. :)

Reply:
Weighting issues - sub-optimal performance vs actual bug I'd say that keeping all fields still wins. But yeah, spec issue is the proper place to discuss this.

Reply:
@kubawach please provide a permalink to your example of TracingRequestStreamHandler, I couldn't find it.

Reply:
@yurishkuro here you go: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/aws-lambda-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambda/v1_0/ApiGatewayProxyRequest.java

Reply:
@kubawach thanks. If I am reading this correctly, I wouldn't classify this usage as "on the extract path", and I probably would've implemented that check differently in the first place, the use of fields() to determine "noHttpPropagationNeeded" seems a bit round-about. Here's the exact usage link, for reference:

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/1b5df6d78a6877dab38f573b44d0358927d1ab3d/instrumentation/aws-lambda-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambda/v1_0/ApiGatewayProxyRequest.java#L27